PLFM-4731 - Error sending messages to Synapse users re: character limitations

hhu94 opened the pull request.

Commit:
Change to, cc, bcc types from String to byte[]

Commit:
Add non-latin1 tests for to, cc and bcc

Commit:
Change error message for blob size

Commit:
Add backup object and implement translation

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/MessageUtils.java:
```diff
@@ -99,9 +99,19 @@ public static void copyDBOToDTO(DBOMessageToUser info, MessageToUser bundle) {
-		bundle.setCc(info.getCc());
-		bundle.setBcc(info.getBcc());
+		try {
+			if (info.getTo() != null) {
```
what if info.getTo()==null?  Should this code then call bundle.setTo(null)?

hhu94 replied:
I made the assumption that the bundle passed in had all null fields, but you're right, it's better to explicitly set the field to null.

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java:
```diff
@@ -0,0 +1,231 @@
+package org.sagebionetworks.repo.model.dbo.persistence;
```
Please open a Jira issue to remove DBOMessageToUserBackup in the subsequent stack. (Assuming elimination of the class is part of the strategy.)

hhu94 replied:
Tracking that task here https://sagebionetworks.jira.com/browse/PLFM-4735

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java:
```diff
@@ -214,23 +226,71 @@ public MigrationType getMigratableTableType() {
 			@Override
-			public DBOMessageToUser createBackupFromDatabaseObject(DBOMessageToUser dbo) {
-				return dbo;
+			public DBOMessageToUserBackup createBackupFromDatabaseObject(DBOMessageToUser dbo) {
```
This doesn't help us get to the state in which the backup is DBOMessageToUser.  Is your plan that DBOMessageToUserBackup would become a permanent part of the migration scheme?

Commit:
Set null explicitly during DBO/DTO translation

Commit:
Update translation logic

The new logic enables us to remove the backup object after the migration
is complete. This cleanup can be done any time, as migration will work
properly even without it.

Commit:
Undo accidental string replacement in comment

Commit:
Synchronize list modified by multiple threads

Commit:
Undo synchronized list and fix concurrency bug

The list synchronized by the previous commit wasn't the source of the
bug. This commit fixes the bug by making sure no list is shared between
threads.

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUser.java:
```diff
@@ -247,8 +318,8 @@ public DBOMessageToUser createBackupFromDatabaseObject(DBOMessageToUser dbo) {
 		int result = 1;
-		result = prime * result + ((bcc == null) ? 0 : bcc.hashCode());
-		result = prime * result + ((cc == null) ? 0 : cc.hashCode());
+		result = prime * result + Arrays.hashCode(bytesBcc);
```
Please let your IDE auto-generate hashcode, equals, toString (if applicable).

brucehoff commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserBackup.java:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	@Override
+	public int hashCode() {
```
regenerate hashCode and equals using IDE

brucehoff commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java:
```diff
@@ -0,0 +1,89 @@
+		testNonToCcBccFieldsEqual(backup, translated);
+		assertArrayEquals(backup.getTo().getBytes("UTF-8"), translated.getBytesTo());
+		assertArrayEquals(backup.getCc().getBytes("UTF-8"), translated.getBytesCc());
+		assertArrayEquals(backup.getBcc().getBytes("UTF-8"), translated.getBytesBcc());
```
add assertions that the String fields are null.

brucehoff commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java:
```diff
@@ -0,0 +1,89 @@
+		testNonToCcBccFieldsEqual(backup, translated);
+		assertArrayEquals(backup.getBytesTo(), translated.getBytesTo());
+		assertArrayEquals(backup.getBytesCc(), translated.getBytesCc());
+		assertArrayEquals(backup.getBytesBcc(), translated.getBytesBcc());
```
assertions that string fields are null

brucehoff replied:
add test for create backup from database object.

Commit:
Add auto-generated equals, hashCode, toString

For DBOMessageToUser and DBOMessageToUserBackup

Commit:
Add null asserts and test for createBackupFromDBO

brucehoff commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java:
```diff
@@ -16,48 +17,97 @@
 		assertArrayEquals(backup.getBcc().getBytes("UTF-8"), translated.getBytesBcc());
+		assertNull(backup.getBytesTo());
+		assertNull(backup.getBytesCc());
+		assertNull(backup.getBytesBcc());
```
not necessary

brucehoff commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/persistence/DBOMessageToUserTest.java:
```diff
@@ -16,48 +17,97 @@
+		assertEquals(dbo.getProfileSettingEndpoint(), translated.getProfileSettingEndpoint());
+		assertEquals(dbo.getWithUnsubscribeLink(), translated.getWithUnsubscribeLink());
+		assertEquals(dbo.getWithProfileSettingLink(), translated.getWithProfileSettingLink());
+		assertEquals(dbo.getIsNotificationMessage(), translated.getIsNotificationMessage());
```
check that backup string fields are null

Commit:
Move null assertions to the right test

brucehoff commented:
Looks great!
